src/coreclr/jit/importercalls.cpp

```diff
@@ -6730,6 +6730,17 @@ void Compiler::impMarkInlineCandidate(GenTree*               callNode,
     // candidate, we're done.
     if (call->IsInlineCandidate() || !call->IsGuardedDevirtualizationCandidate())
     {
+        // Runtime lookup is expected to be spilled into a temp for inline candidates.
+        CallArg* instParam = call->gtArgs.FindWellKnownArg(WellKnownArg::InstParam);
+        if ((instParam != nullptr) && instParam->GetNode()->OperIs(GT_RUNTIMELOOKUP))
+        {
+            const unsigned lookupTmp = lvaGrabTemp(true DEBUGARG("spilling runtimelookup"));
+            impStoreTemp(lookupTmp, instParam->GetNode(), CHECK_SPILL_NONE);
+            GenTreeLclVar* lcl = gtNewLclvNode(lookupTmp, TYP_I_IMPL);
+            lcl->gtFlags |= GTF_VAR_CONTEXT;
+            instParam->SetEarlyNode(lcl);
+        }
```

Comment:
Is it ok to reorder its evaluation with other arguments?

Reply:
I mean both the IL stack but also the other operands of the call. Those will get reordered with the runtime lookup.

The reordering might still be illegal even if the runtime lookup is invariant as a whole, for example if runtime lookups can throw exceptions (which presumably they can if they result in type loading?).

Reply:
I'll check how diffs look like with proper spilling

Reply:
Outside x86 the instantiation parameter is inserted after "this" but before the rest of the arguments, so if the runtime lookup can throw exceptions it seems like we're already getting it wrong today.

Reply:
I think it's similar to how we handle OOM side-effects from allocations

Reply:
Yes, the runtime lookup can throw exceptions for invalid programs or for out-of-memory situations. We do not guarantee the exact position where the exception is going to be thrown.